sqlglot/optimizer/annotate_types.py

```diff
@@ -97,22 +104,54 @@ class TypeAnnotator:
         },
     }
 
-    def __init__(self, schema=None, annotators=None, coerces_to=None):
+    TRAVERSABLES = {exp.Select, exp.Union, exp.UDTF, exp.Subquery}
+
+    def __init__(self, schema=None, annotators=None, coerces_to=None, type_mapping=None):
         self.schema = schema
         self.annotators = annotators or self.ANNOTATORS
         self.coerces_to = coerces_to or self.COERCES_TO
+        self.type_mapping = type_mapping or TYPE_MAPPING
 
     def annotate(self, expression):
+        scopes = traverse_scope(expression) if expression.__class__ in self.TRAVERSABLES else []
+
+        for scope in scopes:
+            # First annotate the current expression's columns references
+            for col in scope.columns:
```

Comment:
I don't think `_annotate_args` stops at subscopes - Will we be passing over the subscopes multiple times?

Reply:
Yes, we need to somehow mark expressions as "processed", otherwise they'll be visited multiple times. Note that this mainly applies to expressions like SELECT, WITH, WHERE etc that don't have a type. After we annotate an expression we don't explore its subtree again, because there's that `if expression.type: return` check in `_annotate`.

Reply:
I can try the "sentinel type" approach so that select.type doesn't yield None if we've visited that select expression in the past and we can prune anyway.

Reply:
Ah, gotcha

Reply:
how about None, DataType.UNKNOWN

Reply:
Here's how scope itself prunes the tree: https://github.com/tobymao/sqlglot/blob/main/sqlglot/optimizer/scope.py#L522

Reply:
But I think it's even easier in your case, as you can just pass in a prune callback that checks for the `type` annotation

Reply:
I think the `set_type` flag can solve the original issue discussed here. Let me know if I missed something.